compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/AbstractCodeGenerator.kt

```diff
@@ -0,0 +1,65 @@
+package edu.cornell.cs.apl.viaduct.codegeneration
+
+import com.squareup.kotlinpoet.CodeBlock
+import edu.cornell.cs.apl.viaduct.errors.IllegalInternalCommunicationError
+import edu.cornell.cs.apl.viaduct.syntax.Protocol
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.AtomicExpressionNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.DeclarationNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.LetNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.OutParameterInitializationNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.OutputNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.SendNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.SimpleStatementNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.UpdateNode
+
+abstract class AbstractCodeGenerator : CodeGenerator {
+
+    override fun simpleStatement(protocol: Protocol, stmt: SimpleStatementNode): CodeBlock {
+        return when (stmt) {
+            is LetNode -> let(protocol, stmt)
+
+            is DeclarationNode -> declaration(protocol, stmt)
+
+            is UpdateNode -> update(protocol, stmt)
+
+            is OutParameterInitializationNode -> outParameterInitialization(protocol, stmt)
+
+            is OutputNode -> output(protocol, stmt)
+
+            is SendNode -> throw IllegalInternalCommunicationError(stmt)
+        }
+    }
+
+    abstract fun let(protocol: Protocol, stmt: LetNode): CodeBlock
+
+    abstract fun declaration(protocol: Protocol, stmt: DeclarationNode): CodeBlock
+
+    abstract fun update(protocol: Protocol, stmt: UpdateNode): CodeBlock
+
+    abstract fun outParameterInitialization(protocol: Protocol, stmt: OutParameterInitializationNode): CodeBlock
+
+    abstract fun output(protocol: Protocol, stmt: OutputNode): CodeBlock
+}
+
+abstract class SingleProtocolCodeGenerator : AbstractCodeGenerator() {
```

Comment:
I don't think we will ever use this.

Reply:
This mirrors `SingleProtocolInterpreter` which is used by ZKP and commitment back ends